### PR TITLE
Transparent Web Entities

### PR DESCRIPTION
**Pull request by kasenvr**
Apparently I thought it would be hard all along, a bit of Googling simply shows that Qt sucks at documentation and that the #1 source of information for Qt is 3rd party websites and bug reports.

References:

https://bugreports.qt.io/browse/QTBUG-29040
https://stackoverflow.com/questions/19549975/setting-transparent-background-for-qwebview
https://bugreports.qt.io/browse/QTBUG-43228

Either way, it's a two line change but I'm not sure I'm doing it in the best way?

Caveats:
You have to lower the opacity below 1.0 on the entity, actually it has to be below .99 specifically, it can't be .990 - 1.00.

Also this applies to all web entities and so the preferred test site for this: https://google.com/ will have its background removed, which is cool for testing but I imagine this will happen to most other websites as well so ultimately, there should be a toggle to engage this as its kind of hacking into the page just like the EventBridge does for every page.

tl;dr this is simply to prove it can be done and Qt had the power all along, just gotta figure out how to proceed. :)

### CODE REVIEW

**Comment by kasenvr** [comment]
https://github.com/axln/radial-menu-js Madders pointed this out to me when I asked about radial menus, might be good inspiration.

But we all know how I like things implemented... https://github.com/axln/radial-menu-vue ;)

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (macOS-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR605/Vircadia-Alpha-PR605-d333955_1601ea92.dmg

**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR605/Vircadia-Alpha-PR605-d333955.tar_1601ea92.Z
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR605/Vircadia-Alpha-PR605-d333955_1601ea92.sh
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR605/Vircadia-Alpha-PR605-d333955.tar_1601ea92.gz

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR605/Vircadia-Alpha-PR605-d333955_1601ea92.exe

**Comment by HifiExperiments** [comment]
I’ve always wondered if there’s a way to fix this without changing the QML or adding a new entity property.  can we get the texture we get from Qt to include transparency from the original website?  that way, we can just detect the transparent texture and use the transparent shader automatically

**Comment by kasenvr** [comment]
I didn't know that's how it worked, is a web entity like a local screen share essentially? If so, it would have to be something along the lines of PNG and we would want to capture that somehow to find out. I know how to add a property to QML, not so much how to intercept the rendering but that would be cool, wouldn't it...

It looks like I have to add a property here, does that not mean a protocol bump then if I do that?

**Comment by HifiExperiments** [comment]
discussed the first part in PM but to repeat here, but basically the answer is yes, Qt gives us a texture every frame (at whatever fps you specify).  if that texture passed through the alpha from the webpage, we could detect that easily and render it correctly

otherwise, yes, if you add a property here, you’d need to bump the protocol

**Comment by kasenvr** [comment]
So I've implemented the ability to set the background color from the entity properties, it however is not implemented into the create app (yet), we want to test out what colors work, why or why not... I've filed a bug with Qt to ascertain why the docs do not outline what backgroundColors can be used and how.

We're getting closer to making this an implementable feature. :)

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (macOS-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR605/Vircadia-Alpha-PR605-881a229_43362916.dmg

**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR605/Vircadia-Alpha-PR605-881a229.tar_43362916.Z
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR605/Vircadia-Alpha-PR605-881a229_43362916.sh
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR605/Vircadia-Alpha-PR605-881a229.tar_43362916.gz

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR605/Vircadia-Alpha-PR605-881a229_43362916.exe

**Comment by vegaslon** [comment]
well the opacity value is actually the first two characters of the color not the last 2.
https://bl.ocks.org/lopspower/03fb1cc0ac9f32ef38f4

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (macOS-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR605/Vircadia-Alpha-PR605-a9994a3_ca974504.dmg

**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR605/Vircadia-Alpha-PR605-a9994a3.tar_ca974504.Z
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR605/Vircadia-Alpha-PR605-a9994a3.tar_ca974504.gz
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR605/Vircadia-Alpha-PR605-a9994a3_ca974504.sh

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR605/Vircadia-Alpha-PR605-a9994a3_ca974504.exe

File: interface/resources/qml/Web3DSurface.qml [hunk 1]
```diff
@@ -43,11 +53,12 @@ Item {
             root.item = newItem
             root.item.url = url
             root.item.scriptUrl = scriptUrl
+            root.item.transparentBackground = webBackgroundColor.endsWith("FF") ? false : true
```
Comment (vegaslon) [inline interface/resources/qml/Web3DSurface.qml:56]: Web background is always transparent unless the background has full blue in it?

Comment (kasenvr) [reply interface/resources/qml/Web3DSurface.qml:56]: Haha, that's exactly the line of code I was going to ask about. Soooo... if the opacity is at the beginning, then that's wrong. If the opacity is at the end, then that's correct. How about I update it and see what results you get?

I use the console to test this PR. "editEntity"

**Comment by kasenvr** [comment]
image
This results in this
image

File: interface/resources/qml/Web3DSurface.qml (hunk 1 above)
Comment (vegaslon) [reply interface/resources/qml/Web3DSurface.qml:56]: Yep that would be great. Even this shows transparency being first. https://doc.qt.io/qt-5/qml-color.html

Comment (kasenvr) [reply interface/resources/qml/Web3DSurface.qml:56]: Indeed it does, I was following that, but when I follow that convention... it doesn't work like I expected! Maybe I'm doing it wrong (again..)

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (macOS-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR605/Vircadia-Alpha-PR605-5fe8871_7e5e513d.dmg

**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR605/Vircadia-Alpha-PR605-5fe8871.tar_7e5e513d.Z
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR605/Vircadia-Alpha-PR605-5fe8871.tar_7e5e513d.gz
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR605/Vircadia-Alpha-PR605-5fe8871_7e5e513d.sh

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR605/Vircadia-Alpha-PR605-5fe8871_7e5e513d.exe

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (macOS-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR605/Vircadia-Alpha-PR605-24aeae3_8d904d39.dmg

**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR605/Vircadia-Alpha-PR605-24aeae3.tar_8d904d39.gz
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR605/Vircadia-Alpha-PR605-24aeae3.tar_8d904d39.Z
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR605/Vircadia-Alpha-PR605-24aeae3_8d904d39.sh

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR605/Vircadia-Alpha-PR605-24aeae3_8d904d39.exe

File: interface/resources/qml/controls/+webengine/FlickableWebViewCore.qml [hunk 2]
```diff
@@ -98,6 +99,7 @@ Item {
 
         width: parent.width
         height: parent.height
+        backgroundColor: (flick.useBackground) ? "white" : "transparent"
```
[the remark this answers is not in the document]
Comment (kasenvr) [reply interface/resources/qml/controls/+webengine/FlickableWebViewCore.qml:102]: I tried using {} per standard QML conventions, but apparently that threw errors saying something along the lines of "Cannot convert QJSValue to QColor."

However, it appears from the docs https://doc.qt.io/qt-5/qml-qtwebengine-webengineview.html#backgroundColor-prop that the default value was white anyway. So I think it would have ended up being that way anyway? I sincerely hope so.

[the remark this answers is not in the document]
Comment (kasenvr) [reply interface/resources/qml/controls/+webengine/FlickableWebViewCore.qml:102]: Actually, upon testing it does work, but it does throw that warning/error still.

[the remark this answers is not in the document]
Comment (kasenvr) [reply interface/resources/qml/controls/+webengine/FlickableWebViewCore.qml:102]: According to this beautiful test page
`<!DOCTYPE html>
<html>
<head>
<style>
body {
  background-color: yellow;
}
</style>
</head>
<body>

<h1 style="font-size: 64px; color: white;">This is a test. (:</h1>

</body>
</html>`
Setting it to "white" does not affect the background if there is one already on the body for example. I think the web as a whole defaults to white regardless if you don't specify, so it will never seem out of the ordinary, I imagine.

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR605/Vircadia-Alpha-PR605-6439ecb_103288b8.sh
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR605/Vircadia-Alpha-PR605-6439ecb.tar_103288b8.gz
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR605/Vircadia-Alpha-PR605-6439ecb.tar_103288b8.Z

**build (macOS-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR605/Vircadia-Alpha-PR605-6439ecb_103288b8.dmg

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR605/Vircadia-Alpha-PR605-6439ecb_103288b8.exe

**Comment by vegaslon** [comment]
EntityItemProperties.cpp needs a new jsdoc entry for your new property.

**Comment by kasenvr** [comment]
Thanks for the reminder!

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR605/Vircadia-Alpha-PR605-1b93b11.tar_d4bfce83.Z
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR605/Vircadia-Alpha-PR605-1b93b11.tar_d4bfce83.gz
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR605/Vircadia-Alpha-PR605-1b93b11_d4bfce83.sh

**build (macOS-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR605/Vircadia-Alpha-PR605-1b93b11_d4bfce83.dmg

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR605/Vircadia-Alpha-PR605-1b93b11_d4bfce83.exe

**Comment by kasenvr** [comment]
According to Madders, this works as expected. Need +1 tester to make sure I haven't broken anything else though!

File: libraries/networking/src/NetworkingConstants.h [hunk 3]
```diff
@@ -41,7 +41,15 @@ namespace NetworkingConstants {
...
     const QUrl BUILDS_XML_URL("https://highfidelity.com/builds.xml");
     const QUrl MASTER_BUILDS_XML_URL("https://highfidelity.com/dev-builds.xml");
+    
+    // WebEntity Defaults
+    const QString WEB_ENTITY_DEFAULT_SOURCE_URL = "https://vircadia.com/";
+    
+    const QString DEFAULT_AVATAR_COLLISION_SOUND_URL = "https://hifi-public.s3.amazonaws.com/sounds/Collisions-otherorganic/Body_Hits_Impact.wav";
 
+    // CDN URLs
+    const QString S3_URL = "http://s3.amazonaws.com/hifi-public";
```
[the remark this answers is not in the document]
Comment (kasenvr) [reply libraries/networking/src/NetworkingConstants.h:51]: We'll want to tackle that later, yes. :)

File: libraries/entities/src/EntityItemProperties.cpp [hunk 4]
```diff
@@ -1382,6 +1384,9 @@ EntityPropertyFlags EntityItemProperties::getChangedProperties() const {
  * @property {WebInputMode} inputMode="touch" - The user input mode to use.
  * @property {boolean} showKeyboardFocusHighlight=true - <code>true</code> if the entity is highlighted when it has keyboard
  *     focus, <code>false</code> if it isn't.
+ * @property {boolean} useBackground=true - <code>true</code> if the web entity should have a background,
+ *     <code>false</code> if the web entity's background should be transparent. The webpage must have CSS properties for
+ *     <code>background-color: "transparent"</code> set appropriately for this property to have an effect.
```
[the remark this answers is not in the document]
Comment (kasenvr) [reply libraries/entities/src/EntityItemProperties.cpp:1389]: Madders has confirmed that this is a thing, yes.

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (macOS-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR605/Vircadia-Alpha-PR605-ba55dbd_6b96832d.dmg

**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR605/Vircadia-Alpha-PR605-ba55dbd.tar_6b96832d.Z
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR605/Vircadia-Alpha-PR605-ba55dbd_6b96832d.sh
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR605/Vircadia-Alpha-PR605-ba55dbd.tar_6b96832d.gz

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR605/Vircadia-Alpha-PR605-ba55dbd_6b96832d.exe

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR605/Vircadia-Alpha-PR605-d3c00e8.tar_2ac6cbd8.Z
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR605/Vircadia-Alpha-PR605-d3c00e8.tar_2ac6cbd8.gz
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR605/Vircadia-Alpha-PR605-d3c00e8_2ac6cbd8.sh

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR605/Vircadia-Alpha-PR605-d3c00e8_2ac6cbd8.exe

**build (macOS-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR605/Vircadia-Alpha-PR605-d3c00e8_2ac6cbd8.dmg

**Review by ctrlaltdavid — APPROVED** [review]

**Comment by kasenvr** [comment]
I had tested this (obviously) when developing it + Madders tested it. We'll probably catch any weirdness stragglers in the release-candidate.

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (macOS-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR605/Vircadia-Alpha-PR605-1518649_87d4fddb.dmg

**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR605/Vircadia-Alpha-PR605-1518649_87d4fddb.sh
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR605/Vircadia-Alpha-PR605-1518649.tar_87d4fddb.gz
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR605/Vircadia-Alpha-PR605-1518649.tar_87d4fddb.Z

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR605/Vircadia-Alpha-PR605-1518649_87d4fddb.exe
